server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -516,6 +538,43 @@ public final DynamicTemplate findDynamicTemplate(String fieldName, DynamicTempla
         return null;
     }
 
+    public void postProcessDynamicMappers() {
+        if (indexSettings().getIndexVersionCreated().onOrAfter(DYNAMICALLY_MAP_DENSE_VECTORS_INDEX_VERSION)) {
+            List<Mapper> postProcessedMappers = new ArrayList<>();
+
+            Map<String, Long> dynamicDenseVectorFieldNameToDimCount = dynamicMappers.stream()
+                .filter(subClassObj -> subClassObj instanceof NumberFieldMapper)
+                .map(NumberFieldMapper.class::cast)
+                .filter(m -> "float".equals(m.typeName()) && fieldAppliedFromTemplate(m.name()) == false)
+                .collect(Collectors.groupingBy(FieldMapper::name, Collectors.counting()))
+                .entrySet()
+                .stream()
+                .filter(e -> e.getValue() >= MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING && e.getValue() <= MAX_DIMS_COUNT)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            if (dynamicDenseVectorFieldNameToDimCount.isEmpty()) {
+                return;
+            }
+
+            for (Mapper mapper : dynamicMappers) {
+                if (dynamicDenseVectorFieldNameToDimCount.containsKey(mapper.name())) {
+                    int size = dynamicDenseVectorFieldNameToDimCount.get(mapper.name()).intValue();
+                    DenseVectorFieldMapper.Builder builder = new DenseVectorFieldMapper.Builder(
+                        mapper.name(),
+                        indexSettings().getIndexVersionCreated(),
+                        size
+                    );
```

Comment:
Do we need this new builder option that accepts size if we allow it to be dynamically set?

It seems better to me to not require this new builder constructor and allow dims to be null in `DenseVectorFieldMapper`.

Then within the parse method you could do something like:

```
    public void parse(DocumentParserContext context) throws IOException {
        if (context.doc().getByKey(fieldType().name()) != null) {
            throw new IllegalArgumentException(
                "Field ["
                    + name()
                    + "] of type ["
                    + typeName()
                    + "] doesn't not support indexing multiple values for the same field in the same document"
            );
        }
        if (Token.VALUE_NULL == context.parser().currentToken()) {
            return;
        }
        if (fieldType().dims == null) {
            int dims = elementType.parseDimensionCount(context);
            DenseVectorFieldType newFieldType = new DenseVectorFieldType(
                    context.buildFullName(name),
                    indexVersionCreated,
                    elementType.getValue(),
                    dims,
                    indexed.getValue(),
                    similarity.getValue(),
                    meta.getValue()
                ),
            Mapper update = new DenseVectorFieldMapper(...);
            context.addDynamicMapper(update);
            // Document will be parsed again with this field after mapping update is applied.
            return;
        }
        Field field = fieldType().indexed ? parseKnnVector(context) : parseBinaryDocValuesVector(context);
        context.doc().addWithKey(fieldType().name(), field);
    }
```
Basically, why not allow `DenseVectorFieldMapper` to have a `null` dims, and then when `parse` is called, set the `dims` to the appropriate number. This way we keep concerns separated.

Reply:
I like this discussion, what do you think about deferring this as part of https://github.com/elastic/elasticsearch/issues/97530?